File: src/scene/RiverScene.java
package scene;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import model.Party;
import model.Person;

import org.newdawn.slick.Animation;
import org.newdawn.slick.Color;
import org.newdawn.slick.GameContainer;
import org.newdawn.slick.Image;
import org.newdawn.slick.SlickException;
import org.newdawn.slick.state.StateBasedGame;

import component.Panel;
import component.Positionable;
import component.Positionable.ReferencePoint;
import component.SegmentedControl;
import component.modal.ComponentModal;
import component.modal.MessageModal;
import component.modal.Modal;
import component.parallax.ParallaxPanel;
import component.parallax.ParallaxComponent;
import component.parallax.ParallaxComponentLoop;
import component.sprite.AnimatingSprite;
import component.sprite.Sprite;
import core.GameDirector;
import core.ConstantStore;
import core.SoundStore;

/**
 * A river crossing scene where the user is given a few choices
 * to cross a river.
 */
public class RiverScene extends Scene {
	public static final SceneID ID = SceneID.RIVER;
	
	private final int NUM_CLOUDS = 5;
	private final int CLOUD_OFFSET = 20;
	private final int CLOUD_DISTANCE_VARIANCE = 10;
	private final int CLOUD_OFFSET_VARIANCE = 10;
	private final int CLOUD_DISTANCE = 300;
	
	private final int MAX_RIVER_DEPTH = 8;
	private final int MAX_TOLL = 1000;
	private final int MIN_TOLL = 100;
	private final int FORD_DANGER_DEPTH = 3;
	private final int CAULK_DANGER_DEPTH = 4;
	private final int RIVER_CROSS_TIME = 3000;
	private final int BRIDGE_CROSS_TIME = 5500;
	

	private ParallaxPanel riverParallaxPanel;
	private ParallaxPanel cloudParallaxPanel;
	
	private AnimatingSprite wagon;
	private AnimatingSprite wagonWheels;
	
	private Sprite bridge;
	
	private SegmentedControl crossingChoicesControl;
	private ComponentModal<SegmentedControl> crossingChoicesModal;
	private MessageModal successModal;
	private Modal nextModal;
	
	private Party party;
	
	private int riverDepth;
	private int tollPrice;
	private int crossTime;
	
	private String successModalMessage;

	private boolean haveWaited;
	private boolean waiting;
	private boolean crossingBridge;
	private boolean crossingRiver;
	private boolean didTakeDamage;
	private boolean ani;
	
	public RiverScene(Party party) {
		this.party = party;
		riverDepth = (int) (Math.random() * MAX_RIVER_DEPTH) + 1;
		tollPrice = (int) (Math.random() * MAX_TOLL - MIN_TOLL) + MIN_TOLL;
		haveWaited = false;
		waiting = false;
		crossTime = 0;
	}
	
	@Override
	public void init(GameContainer container, StateBasedGame game) throws SlickException {
		super.init(container, game);
		
		bridge = new Sprite(container, new Image(ConstantStore.PATH_RIVER + "bridge.png"));
		mainLayer.add(bridge, mainLayer.getPosition(ReferencePoint.CENTERCENTER), ReferencePoint.CENTERCENTER, 100, 100);
		
		//Animating wagon sprites
		Animation wagonAni = new Animation();
		wagonAni.addFrame(new Image(ConstantStore.PATH_RIVER + "wagonriver1.png"), 100);
		wagonAni.addFrame(new Image(ConstantStore.PATH_RIVER + "wagonriver2.png"), 100);
		wagonAni.addFrame(new Image(ConstantStore.PATH_RIVER + "wagonriver3.png"), 100);
		wagon = new AnimatingSprite(container, 50, wagonAni, AnimatingSprite.Direction.LEFT);
		wagonAni = new Animation();
		wagonAni.addFrame(new Image(ConstantStore.PATH_RIVER + "wagonwheels1.png"), 100);
		wagonAni.addFrame(new Image(ConstantStore.PATH_RIVER + "wagonwheels2.png"), 100);
		wagonWheels = new AnimatingSprite(container, 44, wagonAni, AnimatingSprite.Direction.LEFT);
		mainLayer.add(wagon, mainLayer.getPosition(ReferencePoint.CENTERCENTER), Positionable.ReferencePoint.CENTERCENTER, -50, 120);
		mainLayer.add(wagonWheels, mainLayer.getPosition(ReferencePoint.CENTERCENTER), Positionable.ReferencePoint.CENTERCENTER, -50, 120);
		wagonWheels.setVisible(false);
		wagon.setVisible(false);
		


		makeChoiceModal();

		Random random = new Random();
		riverParallaxPanel = new ParallaxPanel(container, container.getWidth(), container.getHeight());
		cloudParallaxPanel = new ParallaxPanel(container, container.getWidth(), container.getHeight());
		ParallaxComponent.MAX_DISTANCE = 600;
		ParallaxComponent water = new ParallaxComponentLoop(container, container.getWidth() + 1, new Image(ConstantStore.PATH_GROUND + "water.png", false, Image.FILTER_LINEAR),1);
		riverParallaxPanel.add(water, backgroundLayer.getPosition(ReferencePoint.BOTTOMLEFT), ReferencePoint.BOTTOMLEFT);
		
		//Clouds
		Image[] cloudImages = new Image[3];
		cloudImages[0] = new Image(ConstantStore.PATH_BKGRND + "cloud_a.png", false, Image.FILTER_NEAREST);
		cloudImages[1] = new Image(ConstantStore.PATH_BKGRND + "cloud_b.png", false, Image.FILTER_NEAREST);
		cloudImages[2] = new Image(ConstantStore.PATH_BKGRND + "cloud_c.png", false, Image.FILTER_NEAREST);
		
		for (int i = 0; i < NUM_CLOUDS; i++) {
			int distance = CLOUD_DISTANCE + random.nextInt(CLOUD_DISTANCE_VARIANCE * 2) - CLOUD_DISTANCE_VARIANCE;
			int cloudImage = random.nextInt(cloudImages.length);
			
			int offset = CLOUD_OFFSET + random.nextInt(CLOUD_OFFSET_VARIANCE * 2) - CLOUD_OFFSET_VARIANCE;
			
			ParallaxComponent cloud = new ParallaxComponent(container, cloudImages[cloudImage], distance, true);
			cloudParallaxPanel.add(cloud, backgroundLayer.getPosition(ReferencePoint.TOPLEFT), ReferencePoint.TOPLEFT, 0, offset);
		}
		
		Panel backgroundColor = new Panel(container, new Color(0x579cdd));
		backgroundLayer.add(backgroundColor, backgroundLayer.getPosition(ReferencePoint.BOTTOMLEFT), ReferencePoint.BOTTOMLEFT);
		backgroundLayer.add(riverParallaxPanel, backgroundLayer.getPosition(ReferencePoint.BOTTOMLEFT), ReferencePoint.BOTTOMLEFT, 0, 200);
		Sprite ground = new Sprite(container, container.getWidth() + 1, new Image(ConstantStore.PATH_GROUND + "riveredge.png", false, Image.FILTER_NEAREST));
		backgroundLayer.add(ground, backgroundLayer.getPosition(ReferencePoint.BOTTOMLEFT), ReferencePoint.BOTTOMLEFT);
		Sprite mountain = new Sprite(container, 800, new Image(ConstantStore.PATH_BKGRND + "hill_b.png", false, Image.FILTER_NEAREST));
		backgroundLayer.add(mountain, backgroundLayer.getPosition(ReferencePoint.TOPRIGHT), ReferencePoint.TOPRIGHT, 0, 50);
		mountain = new Sprite(container, 800, new Image(ConstantStore.PATH_BKGRND + "hill_a.png", false, Image.FILTER_NEAREST));
		backgroundLayer.add(mountain, backgroundLayer.getPosition(ReferencePoint.TOPLEFT), ReferencePoint.TOPLEFT, 0, 50);
		ground = new Sprite(container, container.getWidth() + 1, new Image(ConstantStore.PATH_GROUND + "riveredgetop.png", false, Image.FILTER_NEAREST));
		backgroundLayer.add(ground, backgroundLayer.getPosition(ReferencePoint.TOPLEFT), ReferencePoint.TOPLEFT);
		backgroundLayer.add(cloudParallaxPanel, backgroundLayer.getPosition(ReferencePoint.BOTTOMLEFT), ReferencePoint.BOTTOMLEFT);
	}
	
	
	
	@Override
	public void update(GameContainer container, StateBasedGame game, int delta)
			throws SlickException {
		for (ParallaxComponent sprite : riverParallaxPanel.getParallaxComponents()) {
			sprite.update(delta);
		}
		for (ParallaxComponent sprite : cloudParallaxPanel.getParallaxComponents()) {
			sprite.update(delta);
		}
		if ( !isPaused() ) {
			crossTime += delta;
			if (waiting) {
				if ( crossTime > RIVER_CROSS_TIME ) {
					crossTime = 0;
					waiting = false;
					wagon.setVisible(true);
					showModal(nextModal);
				}
			}
			else {
				wagon.update(container, delta);
				if ( crossingRiver ) {
					ani = !ani;
					if (ani)
						wagon.setLocation(wagon.getX(), wagon.getY() - 1);
					if ( crossTime > RIVER_CROSS_TIME ) {
						if (didTakeDamage) {
							SoundStore.get().playSound("Splash");
						}
						crossTime = 0;
						showModal(nextModal);
					}
				}
				else if (crossingBridge) {
					wagon.setLocation(wagon.getX(), wagon.getY() - 1);
					if ( crossTime > BRIDGE_CROSS_TIME ) {
						crossTime = 0;
						showModal(nextModal);
					}
				}
			}
		}
	}

	@Override
	public void enter(GameContainer container, StateBasedGame game)  {
		super.enter(container, game);
		showModal(crossingChoicesModal);
		wagon.setVisible(true);
		SoundStore.get().loopMusic("River");
	}
	
	@Override
	public int getID() {
		return ID.ordinal();
	}
	
	@Override
	public void dismissModal(Modal modal, int button) {
		super.dismissModal(modal, button);
		if (modal == crossingChoicesModal) {
			int[] choice = crossingChoicesModal.getComponent().getSelection();
			switch ( choice[0] ) {
				case 0 : ford(); break;
				case 1 : caulk(); break;
				case 2 : payToll(); break;
				case 3 : delay(); break;
			}
		}
		else if (modal == successModal) {
			SoundStore.get().stopMusic();
			GameDirector.sharedSceneListener().sceneDidEnd(this);
		}
	}
	
	/**
	 * Construct the modal with all of the river crossing choices.  Disable
	 * choices if they are not available
	 */
	public void makeChoiceModal() {
		String[] choices = {"Ford the river", "Caulk your wagon", "Pay the toll", "Wait for an hour"};
		int[] disabled = {-1};
		if (party.getMoney() < tollPrice && haveWaited)
			disabled = new int[] {2, 3};
		else if (party.getMoney() < tollPrice)
			disabled = new int[] {2};
		else if (haveWaited)
			disabled = new int[] {3};
		crossingChoicesControl = new SegmentedControl(container, 600, 150, 2, 2, 20, true, 1, choices);
		if (disabled[0] != -1)
			crossingChoicesControl.setDisabled(disabled);
		makeChoicesModalMessage();
		crossingChoicesModal = new ComponentModal<SegmentedControl>(container, this, successModalMessage, 1, crossingChoicesControl);
	}
	
	/**
	 * Construct the message for the main choices modal
	 */
	public void makeChoicesModalMessage() {
		successModalMessage = String.format("You've come to a river.  It is %d %s deep.\n" +
			"There is a bridge with a $%d toll, and you have $%d.\nWhat do you want to do?",
			riverDepth, (riverDepth == 1) ? "foot" : "feet", tollPrice, party.getMoney());
	}
	
	/**
	 * The logic for fording a river
	 */
	private void ford() {
		crossingRiver = true;
		if (riverDepth >= FORD_DANGER_DEPTH && Math.random() > 0.5) {
			successModalMessage = "Oh no!  Why would you ford a " + riverDepth +
					" foot deep river?  Your party was damaged";
			String deaths = damage();
			successModalMessage += (deaths.equals("")) ? ", but at least no one died!" :
				" and you lost " + deaths + ".";
		}
		else {
			successModalMessage = "You successfully forded the river!  Your party sighs in" +
					" relief";
		}
		successModal = new MessageModal(container, this, successModalMessage);
		nextModal = successModal;
	}
	
	/**
	 * The logic for caulking and floating across a river
	 */
	private void caulk() {
		crossingRiver = true;
		if (riverDepth >= CAULK_DANGER_DEPTH && Math.random() > 0.5) {
			successModalMessage = "Oh no!  Your caulk didn't hold up and water leaked into " +
					"your wagon.  Your party was damaged";
			String deaths = damage();
			successModalMessage += (deaths.equals("")) ? ", but at least no one died!" :
				" and you lost " + deaths + ".";
		}
		else {
			successModalMessage = "Water started seeping into your wagon just as you " +
					"reached the shore, but you make it!  You dump the water " +
					"out of your boots, take a big swill of whiskey, and get " +
					"back on the trail.";
		}
		successModal = new MessageModal(container, this, successModalMessage);
		nextModal = successModal;
	}
	
	/**
	 * The logic for caulking and floating across a river
	 */
	private void payToll() {
		crossingBridge = true;
		party.setMoney(party.getMoney() - tollPrice);
		wagon.setVisible(false);
		wagonWheels.setVisible(true);
		wagon = wagonWheels;
		wagon.setLocation(wagon.getX() + 150, 576);
		SoundStore.get().playSound("RK");
		successModalMessage = "Your party decided to take the easy way out and " +
				"pay the the bridge toll.  Your party members thank you.";
		successModal = new MessageModal(container, this, successModalMessage);
		nextModal = successModal;
	}
	
	/**
	 * The logic for waiting to see if the river goes down
	 */
	private void delay() {
		haveWaited = true;
		waiting = true;
		wagon.setVisible(false);
		riverDepth = (int) (Math.random() * MAX_RIVER_DEPTH) + 1;
		makeChoiceModal();
		nextModal = crossingChoicesModal;
	}

	/**
	 * Damage all party members by a random amount.
	 * @return A String representation of the deaths caused by damage.
	 */
	public String damage() {
		ArrayList<String> deadMembers = new ArrayList<String>();
		didTakeDamage = true;
		if (party.getVehicle() != null) {
			party.getVehicle().decreaseStatus((int)(Math.random()*100));
		}
		List<Person> partyMembers = party.getPartyMembers();
		for ( int i = partyMembers.size() - 1; i >= 0; i-- ) {
			Person person = partyMembers.get(i);
			boolean stillAlive = party.decreaseHealth(person,(int)(Math.random()*100));
			if ( !stillAlive ) {
				deadMembers.add(person.getName());
			}
		}
		
		String deaths = "";
		if ( deadMembers.size() == 1 )
			deaths += deadMembers.get(0);
		else if (deadMembers.size() == 2)
			deaths += deadMembers.get(0) + " and " + deadMembers.get(1);
		else if (deadMembers.size() > 2) {
			for (int i = 0; i < deadMembers.size(); i++) {
				if ( i == deadMembers.size() -1 )
					deaths += "and ";
				deaths += deadMembers.get(i);
				if ( i != deadMembers.size() - 1)
					deaths += ", ";
			}
		}
		return deaths;
	}
}

File: web/src/scene/RiverScene.ts
import { Scene } from './Scene';
import { SceneID } from './SceneID';
import { Party } from '../model/Party';
import { Panel } from '../component/Panel';
import { Color } from '../core/Color';
import { ImageStore } from '../core/ImageStore';
import { SoundStore } from '../core/SoundStore';
import { Sprite } from '../component/Sprite';
import { AnimatingSprite } from '../component/sprite/AnimatingSprite';
import { ParallaxPanel } from '../component/parallax/ParallaxPanel';
import { ParallaxComponent } from '../component/parallax/ParallaxComponent';
import { ParallaxComponentLoop } from '../component/parallax/ParallaxComponentLoop';
import { ComponentModal } from '../component/modal/ComponentModal';
import { MessageModal } from '../component/modal/MessageModal';
import { SegmentedControl } from '../component/SegmentedControl';
import { ReferencePoint } from '../component/Component';
import { Component } from '../component/Component';

/**
 * River crossing scene where the player chooses how to cross:
 * Ford, Caulk and float, Pay ferry toll, or Wait for conditions to improve.
 */
export class RiverScene extends Scene {
  private static readonly NUM_CLOUDS = 5;
  private static readonly CLOUD_OFFSET = 20;
  private static readonly CLOUD_DISTANCE_VARIANCE = 10;
  private static readonly CLOUD_OFFSET_VARIANCE = 10;
  private static readonly CLOUD_DISTANCE = 300;

  private static readonly MAX_RIVER_DEPTH = 8;
  private static readonly FORD_DANGER_DEPTH = 3;
  private static readonly CAULK_DANGER_DEPTH = 4;
  private static readonly RIVER_CROSS_TIME = 3000;
  private static readonly BRIDGE_CROSS_TIME = 5500;

  private party: Party;
  private canvasWidth: number;
  private canvasHeight: number;

  private riverParallaxPanel!: ParallaxPanel;
  private cloudParallaxPanel!: ParallaxPanel;

  private wagon!: AnimatingSprite;
  private wagonWheels!: AnimatingSprite;
  private bridge!: Sprite;

  private crossingChoicesModal!: ComponentModal<SegmentedControl>;
  private successModal!: MessageModal;
  private nextModal: ComponentModal<SegmentedControl> | MessageModal | null = null;

  private riverDepth: number;
  private tollPrice: number;
  private crossTime: number = 0;

  private haveWaited: boolean = false;
  private waiting: boolean = false;
  private crossingBridge: boolean = false;
  private crossingRiver: boolean = false;
  private didTakeDamage: boolean = false;
  private bobbingUp: boolean = true;

  constructor(party: Party, canvasWidth: number, canvasHeight: number) {
    super();
    this.party = party;
    this.canvasWidth = canvasWidth;
    this.canvasHeight = canvasHeight;

    this.riverDepth = Math.floor(Math.random() * RiverScene.MAX_RIVER_DEPTH) + 1;
    this.tollPrice = (Math.floor(Math.random() * 20) + 1) * (party.getLocation().getRank() + 1);
  }

  init(): void {

    // Sky background
    const skyColor = new Color(0x57 / 255, 0x9c / 255, 0xdd / 255);
    const sky = new Panel(this.canvasWidth, this.canvasHeight, skyColor);
    this.backgroundLayer.add(sky);

    // Cloud parallax
    this.cloudParallaxPanel = new ParallaxPanel(this.canvasWidth, this.canvasHeight);
    const cloudImages = [
      ImageStore.getImage('CLOUD_A'),
      ImageStore.getImage('CLOUD_B'),
      ImageStore.getImage('CLOUD_C')
    ];

    for (let i = 0; i < RiverScene.NUM_CLOUDS; i++) {
      const distance = RiverScene.CLOUD_DISTANCE +
        Math.floor(Math.random() * RiverScene.CLOUD_DISTANCE_VARIANCE * 2) -
        RiverScene.CLOUD_DISTANCE_VARIANCE;
      const cloudImage = cloudImages[Math.floor(Math.random() * cloudImages.length)];
      const offset = RiverScene.CLOUD_OFFSET +
        Math.floor(Math.random() * RiverScene.CLOUD_OFFSET_VARIANCE * 2) -
        RiverScene.CLOUD_OFFSET_VARIANCE;

      const cloud = new ParallaxComponent(cloudImage, distance, true);
      this.cloudParallaxPanel.add(cloud);
      cloud.setPosition({ x: 0, y: offset }, ReferencePoint.TOPLEFT);
    }
    this.backgroundLayer.add(this.cloudParallaxPanel);

    // Hills
    const hillB = new Sprite(800, undefined, ImageStore.getImage('HILL_B'));
    this.backgroundLayer.add(hillB);
    hillB.setPosition({ x: this.canvasWidth, y: 50 }, ReferencePoint.TOPRIGHT);

    const hillA = new Sprite(800, undefined, ImageStore.getImage('HILL_A'));
    this.backgroundLayer.add(hillA);
    hillA.setPosition({ x: 0, y: 50 }, ReferencePoint.TOPLEFT);

    // River edge top
    const riverEdgeTop = new Sprite(this.canvasWidth + 1, undefined, ImageStore.getImage('RIVER_EDGE_TOP'));
    this.backgroundLayer.add(riverEdgeTop);
    riverEdgeTop.setPosition({ x: 0, y: 0 }, ReferencePoint.TOPLEFT);

    // River parallax (water)
    this.riverParallaxPanel = new ParallaxPanel(this.canvasWidth, this.canvasHeight);
    const water = new ParallaxComponentLoop(this.canvasWidth + 1, ImageStore.getImage('WATER'), 1);
    this.riverParallaxPanel.add(water);
    water.setPosition({ x: 0, y: this.canvasHeight - 200 }, ReferencePoint.BOTTOMLEFT);
    this.backgroundLayer.add(this.riverParallaxPanel);

    // River edge
    const riverEdge = new Sprite(this.canvasWidth + 1, undefined, ImageStore.getImage('RIVER_EDGE'));
    this.backgroundLayer.add(riverEdge);
    riverEdge.setPosition({ x: 0, y: this.canvasHeight }, ReferencePoint.BOTTOMLEFT);

    // Bridge
    this.bridge = new Sprite(200, undefined, ImageStore.getImage('BRIDGE'));
    this.mainLayer.add(this.bridge);
    this.bridge.setPosition(
      { x: this.canvasWidth / 2, y: this.canvasHeight / 2 },
      ReferencePoint.CENTERCENTER,
      100,
      100
    );

    // Wagon animation (3 frames)
    const wagonFrames = [
      ImageStore.getImage('WAGON_RIVER_1'),
      ImageStore.getImage('WAGON_RIVER_2'),
      ImageStore.getImage('WAGON_RIVER_3')
    ];
    this.wagon = new AnimatingSprite(wagonFrames, 100, 'left');
    this.wagon.setSpeed(50);
    this.mainLayer.add(this.wagon);
    this.wagon.setPosition(
      { x: this.canvasWidth / 2, y: this.canvasHeight / 2 },
      ReferencePoint.CENTERCENTER,
      -50,
      120
    );
    this.wagon.setVisible(false);

    // Wagon wheels animation (2 frames)
    const wheelFrames = [
      ImageStore.getImage('WAGON_WHEELS_1'),
      ImageStore.getImage('WAGON_WHEELS_2')
    ];
    this.wagonWheels = new AnimatingSprite(wheelFrames, 100, 'left');
    this.wagonWheels.setSpeed(44);
    this.mainLayer.add(this.wagonWheels);
    this.wagonWheels.setPosition(
      { x: this.canvasWidth / 2, y: this.canvasHeight / 2 },
      ReferencePoint.CENTERCENTER,
      -50,
      120
    );
    this.wagonWheels.setVisible(false);

    // Initial choice modal
    this.makeChoiceModal();
  }

  override enter(): void {
    super.enter();
    this.showModal(this.crossingChoicesModal);
    this.wagon.setVisible(true);
    SoundStore.loopMusic('River');
  }

  override update(delta: number): void {
    super.update(delta);

    // Update parallax
    for (const component of this.riverParallaxPanel.getComponents()) {
      if (component instanceof ParallaxComponent) {
        component.update(delta);
      }
    }
    for (const component of this.cloudParallaxPanel.getComponents()) {
      if (component instanceof ParallaxComponent) {
        component.update(delta);
      }
    }

    if (!this.isPaused()) {
      this.crossTime += delta;

      if (this.waiting) {
        if (this.crossTime > RiverScene.RIVER_CROSS_TIME) {
          this.crossTime = 0;
          this.waiting = false;
          this.wagon.setVisible(true);
          if (this.nextModal) {
            this.showModal(this.nextModal);
          }
        }
      } else {
        this.wagon.update(delta);

        if (this.crossingRiver) {
          // Bobbing animation
          if (this.bobbingUp) {
            this.wagon.setPosition(
              { x: this.wagon.getX(), y: this.wagon.getY() - 1 },
              ReferencePoint.TOPLEFT
            );
          }
          this.bobbingUp = !this.bobbingUp;

          if (this.crossTime > RiverScene.RIVER_CROSS_TIME) {
            if (this.didTakeDamage) {
              SoundStore.playSound('Splash');
            }
            this.crossTime = 0;
            if (this.nextModal) {
              this.showModal(this.nextModal);
            }
          }
        } else if (this.crossingBridge) {
          this.wagon.setPosition(
            { x: this.wagon.getX(), y: this.wagon.getY() - 1 },
            ReferencePoint.TOPLEFT
          );

          if (this.crossTime > RiverScene.BRIDGE_CROSS_TIME) {
            this.crossTime = 0;
            if (this.nextModal) {
              this.showModal(this.nextModal);
            }
          }
        }
      }
    }
  }

  override dismissModal(modal: Component, button: number): void {
    super.dismissModal(modal, button);

    const currentModal = modal;

    if (currentModal === this.crossingChoicesModal) {
      const choice = this.crossingChoicesModal.getComponent().getSelection();
      if (choice.length > 0) {
        switch (choice[0]) {
          case 0: this.ford(); break;
          case 1: this.caulk(); break;
          case 2: this.payToll(); break;
          case 3: this.delay(); break;
        }
      }
    } else if (currentModal === this.successModal) {
      SoundStore.stopMusic();
      this.exit();
    }
  }

  getID(): SceneID {
    return SceneID.RIVER;
  }

  /**
   * Create the modal with crossing choices
   */
  private makeChoiceModal(): void {
    const choices = ['Ford the river', 'Caulk your wagon', 'Pay the toll', 'Wait for an hour'];
    const disabled: number[] = [];

    if (this.party.getMoney() < this.tollPrice) {
      disabled.push(2);
    }
    if (this.haveWaited) {
      disabled.push(3);
    }

    const control = new SegmentedControl(
      600,
      150,
      2,
      2,
      20,
      true,
      1,
      choices
    );

    if (disabled.length > 0) {
      control.setDisabled(disabled);
    }

    const message = `You've come to a river. It is ${this.riverDepth} ${this.riverDepth === 1 ? 'foot' : 'feet'} deep.\nThere is a bridge with a $${this.tollPrice} toll, and you have $${this.party.getMoney()}.\nWhat do you want to do?`;

    this.crossingChoicesModal = new ComponentModal(message, 1, control);
  }

  /**
   * Ford the river (wade across)
   */
  private ford(): void {
    this.crossingRiver = true;

    if (this.riverDepth >= RiverScene.FORD_DANGER_DEPTH && Math.random() > 0.5) {
      const deaths = this.damage();
      this.successModal = new MessageModal(
        `Oh no! Why would you ford a ${this.riverDepth} foot deep river? Your party was damaged` +
        (deaths === '' ? ', but at least no one died!' : ` and you lost ${deaths}.`)
      );
    } else {
      this.successModal = new MessageModal(
        'You successfully forded the river! Your party sighs in relief'
      );
    }

    this.nextModal = this.successModal;
  }

  /**
   * Caulk and float across
   */
  private caulk(): void {
    this.crossingRiver = true;

    if (this.riverDepth >= RiverScene.CAULK_DANGER_DEPTH && Math.random() > 0.5) {
      const deaths = this.damage();
      this.successModal = new MessageModal(
        "Oh no! Your caulk didn't hold up and water leaked into your wagon. Your party was damaged" +
        (deaths === '' ? ', but at least no one died!' : ` and you lost ${deaths}.`)
      );
    } else {
      this.successModal = new MessageModal(
        'Water started seeping into your wagon just as you reached the shore, but you make it! ' +
        'You dump the water out of your boots, take a big swill of whiskey, and get back on the trail.'
      );
    }

    this.nextModal = this.successModal;
  }

  /**
   * Pay the toll and cross the bridge
   */
  private payToll(): void {
    this.crossingBridge = true;
    this.party.setMoney(this.party.getMoney() - this.tollPrice);

    // Switch to wheel sprite
    this.wagon.setVisible(false);
    this.wagonWheels.setVisible(true);
    this.wagon = this.wagonWheels;
    this.wagon.setPosition(
      { x: this.wagon.getX() + 150, y: 576 },
      ReferencePoint.TOPLEFT
    );

    SoundStore.playSound('RK');

    this.successModal = new MessageModal(
      'Your party decided to take the easy way out and pay the bridge toll. Your party members thank you.'
    );

    this.nextModal = this.successModal;
  }

  /**
   * Wait to see if conditions improve
   */
  private delay(): void {
    this.haveWaited = true;
    this.waiting = true;
    this.wagon.setVisible(false);

    // Random new depth
    this.riverDepth = Math.floor(Math.random() * RiverScene.MAX_RIVER_DEPTH) + 1;

    this.makeChoiceModal();
    this.nextModal = this.crossingChoicesModal;
  }

  /**
   * Apply damage to party and vehicle
   * @returns String describing who died
   */
  private damage(): string {
    const deadMembers: string[] = [];
    this.didTakeDamage = true;

    // Damage vehicle
    const vehicle = this.party.getVehicle();
    if (vehicle) {
      vehicle.decreaseStatus(Math.floor(Math.random() * 100));
    }

    // Damage party members
    const partyMembers = this.party.getPartyMembers();
    for (let i = partyMembers.length - 1; i >= 0; i--) {
      const person = partyMembers[i];
      const stillAlive = this.party.decreaseHealth(person, Math.floor(Math.random() * 100));
      if (!stillAlive) {
        deadMembers.push(person.getName());
      }
    }

    // Format death list
    if (deadMembers.length === 0) {
      return '';
    } else if (deadMembers.length === 1) {
      return deadMembers[0];
    } else if (deadMembers.length === 2) {
      return `${deadMembers[0]} and ${deadMembers[1]}`;
    } else {
      let deaths = '';
      for (let i = 0; i < deadMembers.length; i++) {
        if (i === deadMembers.length - 1) {
          deaths += 'and ';
        }
        deaths += deadMembers[i];
        if (i !== deadMembers.length - 1) {
          deaths += ', ';
        }
      }
      return deaths;
    }
  }
}
